src/Framework/Receipts/Actions/GenerateConfirmationPageReceipt.php

```diff
@@ -46,26 +49,56 @@ protected function getCustomFields(Donation $donation): array
         });
 
         $receiptDetails = [];
-        foreach($customFields as $field) {
+        foreach ($customFields as $field) {
             /** @var Field|HasLabel|HasName $field */
-            if ($field->shouldStoreAsDonorMeta()) {
-                if (!metadata_exists('donor', $donation->donor->id, $field->getName()) ) {
+            if ($field->hasReceiptValue()) {
+                try {
+                    $value = $field->isReceiptValueCallback() ? $field->getReceiptValue()($field, $donation) : null;
+                } catch (Exception $e) {
+                    $value = null;
+
+                    Log::error('Error getting receipt value for field', [
+                        'field' => $field->getName(),
+                        'error' => $e->getMessage(),
+                        'trace' => $e->getTraceAsString(),
+                    ]);
+                }
+            } elseif ($field->getScope()->isDonor()) {
+                if (!metadata_exists('donor', $donation->donor->id, $field->getName())) {
                     continue;
                 }
 
                 $value = give()->donor_meta->get_meta($donation->donor->id, $field->getName(), true);
-            } else {
-                if (!metadata_exists('donation', $donation->id, $field->getName()) ) {
+            } elseif ($field->getScope()->isDonation()) {
+                if (!metadata_exists('donation', $donation->id, $field->getName())) {
                     continue;
                 }
 
                 $value = give()->payment_meta->get_meta($donation->id, $field->getName(), true);
+            } else {
+                $value = null;
             }
 
-            $receiptDetails[] = new ReceiptDetail(
-                $field->getLabel(),
-                $value
+            $value = apply_filters(
+                sprintf("givewp_donation_confirmation_page_field_value_for_%s", $field->getName()),
+                $value,
+                $field,
+                $donation
+            );
+
+            $label = apply_filters(
+                sprintf("givewp_donation_confirmation_page_field_label_for_%s", $field->getName()),
+                $field->hasReceiptLabel() ? $field->getReceiptLabel() : $field->getLabel(),
+                $field,
+                $donation
             );
+
+            if (!empty($label)){
+                 $receiptDetails[] = new ReceiptDetail(
+                    $label,
+                    $value ?? ''
```

Comment:
What is the use case for adding a label to the Receipt, but not a value?

Reply:
It's a good question, I think it has to do with the option to _show in receipt_ - where the expectation is it will show up as a line item - not necessarily meaning it has a value.

Reply:
Is this basically just handling scenarios where the value is null but we want the value to always be a string, or something? I'm confused, too. Don't all fields have a corresponding value?

Reply:
Circling back on this for more context.  This actually comes down to a UX decision - when a form has fields that are not required, should they still show up in the receipt as label & empty value?  This is currently how it works but we can talk about this as a team.

Reply:
Update: via Slack we decided to maintain the status quo and possibly revisit if necessary.